### PR TITLE
Support for fixed percentage

### PR DESCRIPTION
**Pull request by worldtiki**
Added a new attack mode:
- fixed (existing): kills a specific number of pods
- random-max-percentage (existing): kills a random percentage of pods
- fixed-percentage (new): kills a fixed percentage of pods

Also fixed the bug mentioned here: https://github.com/asobti/kube-monkey/issues/81

### CODE REVIEW

File: victims/victims.go [hunk 1]
```diff
@@ -126,6 +129,42 @@ func (v *VictimBase) DeletePod(clientset kube.Interface, podName string) error {
 	return clientset.CoreV1().Pods(v.namespace).Delete(podName, deleteopts)
 }
 
+// Removes a fixed percentage of pods for the victim
+func (v *VictimBase) DeletePodsFixedPercentage(clientset kube.Interface, killPercentage int) error {
```
Comment (asobti) [inline victims/victims.go]: Can you add validation that `killValue` has a valid value (0-100) and a corresponding unit test.

While you're at that, can you add the same validation to `DeleteRandomPodsMaxPercentage` too. Thanks!

File: chaos/chaos.go [hunk 2]
```diff
@@ -120,9 +119,10 @@ func (c *Chaos) terminate(clientset kube.Interface) error {
 	switch killType {
 	case config.KillFixedLabelValue:
 		return c.Victim().DeleteRandomPods(clientset, killValue)
```
[the remark this answers is not in the document]
Comment (worldtiki) [reply chaos/chaos.go]: Thanks for the feedback. I agree with your suggestion, I'll update the pull request.

File: victims/victims.go [hunk 3]
```diff
@@ -126,6 +129,60 @@ func (v *VictimBase) DeletePod(clientset kube.Interface, podName string) error {
...
+	if killPercentage < 0 || killPercentage > 100 {
```
[the remark this answers is not in the document]
Comment (worldtiki) [reply victims/victims.go]: I think these are two distinct cases. The number of pods is dynamic, so it's acceptable to allow users to configure any value of pods to kill when using the "fixed number" kill mode.
However, a percentage over 100 is a configuration error.

I think a better approach would be to move this check upstream (when loading the labels with these values) and failing there (that's a bit out of scope of these changes and should probably be done in a separate pull request).

File: victims/victims.go [hunk 4]
```diff
@@ -126,6 +129,60 @@ func (v *VictimBase) DeletePod(clientset kube.Interface, podName string) error {
...
+		return fmt.Errorf("The kill percentage needs to be between 0 and 100. It was %d for %s %s", killPercentage, v.kind, v.name)
```
[the remark this answers is not in the document]
Comment (worldtiki) [reply victims/victims.go]: Thanks. I'll change the message.

File: victims/victims.go [hunk 5]
```diff
@@ -126,6 +129,60 @@ func (v *VictimBase) DeletePod(clientset kube.Interface, podName string) error {
...
+
+	pods, err := v.RunningPods(clientset)
+	if err != nil {
+		return err
+	}
+
+	numPods := len(pods)
+
+	r := rand.New(rand.NewSource(time.Now().UnixNano()))
+	killPercentage := (r.Intn(maxPercentage + 1)) // + 1 because Intn works with half open interval [0,n) and we want [0,n]
```
[the remark this answers is not in the document]
Comment (worldtiki) [reply victims/victims.go]: :)

File: victims/victims.go [hunk 6]
```diff
@@ -126,6 +129,60 @@ func (v *VictimBase) DeletePod(clientset kube.Interface, podName string) error {
...
+	if killPercentage < 0 || killPercentage > 100 {
+		return fmt.Errorf("The kill percentage needs to be between 0 and 100. It was %d for %s %s", killPercentage, v.kind, v.name)
+	}
+	if killPercentage == 0 {
+		glog.V(6).Infof("Not terminating any pods for %s %s as kill percentage is 0\n", v.kind, v.name)
+		// Report success
+		return nil
+	}
+
+	pods, err := v.RunningPods(clientset)
```
[the remark this answers is not in the document]
Comment (worldtiki) [reply victims/victims.go]: I think this is just a general comment and not specific to this pr. No action taken from me here.

File: victims/victims.go [hunk 7]
```diff
@@ -257,3 +241,78 @@ func RandomPodName(pods []v1.Pod) string {
...
+	if killPercentage < 0 {
+		glog.V(6).Infof("Expected kill percentage config %d to be between 0 and 100 for %s %s. Defaulting to 0", killPercentage, v.kind, v.name)
+		killPercentage = 0
+	}
+	if killPercentage > 100 {
+		glog.V(6).Infof("Expected kill percentage config %d to be between 0 and 100 for %s %s. Defaulting to 100", killPercentage, v.kind, v.name)
+		killPercentage = 100
+	}
+
+	pods, err := v.RunningPods(clientset)
```
Comment (asobti) [inline victims/victims.go]: minor comment but the code to get list of all running pods seems to be repeated thrice. Can we move this to a function.

Comment (worldtiki) [reply victims/victims.go]: Pr updated.

**Comment by asobti** [comment]
LGTM. I'll leave it open for @Spellchaser to review as well.

**Comment by asobti** [comment]
I'm going to go ahead and merge this. Thanks for making this change @worldtiki

**Comment by Spellchaser** [comment]
Been busy at conferences/traveling. It addressed my initial comments :)

**Pull request merged**